File: leet-code/diameterOfBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function (root) {
  let ans = 0
  if (root == null) {
    return ans
  }

  let depth = function (root) {
    if (root == null) {
      return 0
    }
    let L = depth(node.left)
    let R = depth(node.right)
    ans = Math.max(ans, L + R + 1) // 左边深度+右边深度+父节点s
    return Math.max(L, R) + 1
  }

  depth(root)
  return ans - 1
}

File: leet-code/diameterOfBinaryTree.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func diameterOfBinaryTree(root *TreeNode) int {
	res := 0
	if root == nil {
		return res
	}
	var depth func(root *TreeNode) int
	depth = func(root *TreeNode) int {
		if root == nil {
			return 0
		}
		L := depth(root.Left)
		R := depth(root.Right)
		res = max(L+R+1, res)
		return max(L, R) + 1
	}
	depth(root)
	return res - 1
}

func max(a int, b int) int {
	if a > b {
		return a
	}
	return b
}
